text/0000-variadic-generics.md

```diff
@@ -0,0 +1,444 @@
+type Tuple<...T, ...U> // invalid
+```
+
+## Implementing `Tuple` `zip`
+[implementing-tuple-zip]: #implementing-tuple-zip
+
+**TODO(dlrobertson)**: It isn't a variadic generics document without a
+implementation of tuple zip :smile:
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+**TODO(dlrobertson)**: Fill this out.
+
+# Prior art
+[prior-art]: #prior-art
+
+## C++1x
+[c++1x]: [#c++1x]
+
+C\+\+11 sets a decent precedent with its variadic templates, which can be
+used to define type-safe variadic functions, among other things. C\+\+11
+has a special case for variadic parameter packs. There is no need to
+introduce a new special case for variadic generics in Rust due to `Tuple`s.
+
+### C++11 Recursion
+[c++11-recursion]: #c++11-recursion
+
+When C\+\+11 introduced parameter packs the typical method used to iterate
+over a list of arguments was to combine overloading and recursion.
+
+```c
+// Implement do_something for types that are expected
+
+// Implement func for nothing (the terminator)
+void func() { }
+
+template<typename T, typename... U> void func(const T& x, const U&&... args) {
+    // Consume x and call do_something on it
+    do_something(x);
+    // Recurse and call func on the rest of the arguments
+    func(args...);
+}
+```
+
+This is a bit complicated and the current implementation of variadic generics
+in Rust will not seek to be compatible with this implementation.
+
+### C++17 Fold syntax
+[c++17-fold-syntax]: #c++17-fold-syntax
+
+C\+\+17 introduced [fold expressions] which is much more similar to the syntax
+that will be implemented, however the current syntax proposed for Rust would
+not allow for the use of binary fold expressions.
+
+```c
+template<typename... Args>
+bool all(Args... args) { return (... && args); }
+```
+
+**TODO(dlrobertson)**: should we follow suit with binary operators? Note this
+may cause issues with a prior syntax that was used for ranges.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Duplicating syntax
+[duplicating-syntax]: #duplicating-syntax
+
+We are using the same syntax (`...` as a prefix) for both declaration and
+expansion. C++ uses `...` as a prefix for the declaration and `...` as a
+postfix for expansion. This does make it much easier to differentiate the
+two. For example, given the following:
+
+```rust
+type Example<...T>
+```
+
+Is `T` a new variadic list of types or did the users intended to unpack
+the `Tuple` of types named `T`? It would be preferable to use `...` as a
+postfix (like C++), but it is currently reserved for ranges.
+
+## Trait bound syntax
+[trait-bound-syntax]: trait-bound-syntax
+
+Is the use of `...T: Trait` a wise use of the syntax. Would it be better
+to define `T` as a non-variadic generic first, followed by the variadic
+definition. For example:
+
+```rust
+fn example<T: Clone, ...T>(...args: T) {
```

Comment:
I would interpret this as first quantifying a single type `T` and then a list of types that are all `T`.

Reply:
This also makes me wonder how you ask for a homogenous tuple of types? `fn min<T, ...U = T>(...args: U)`?